File: 402-remove-k-digits/remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        stack<char> st;
        for(char digit : num){
            while(!st.empty() && k && st.top() > digit){
                st.pop();
                k--;
            }
            st.push(digit);
            
        }
        while(k && !st.empty()){
            st.pop();
            k--;
        }
        string ans="";
        while(!st.empty()){
            ans+=st.top();
            st.pop();
        }
        reverse(ans.begin(), ans.end());
        int i=0;
        while(i<ans.size() && ans[i]=='0') i++;
        ans=ans.substr(i);
        return ans==""?"0":ans;
    }
};

File: 402-remove-k-digits/remove-k-digits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function(num, k) {
    let stack = [];
    for(digit of num){
        while(k && stack.length && stack[stack.length - 1] > digit){
            stack.pop();
            k--;
        }
        stack.push(digit);
    }
        while(k && stack.length){
            stack.pop();
            k--;
        }
        let i=0;
        let ans = stack.join('');
        while(i<ans.length && ans[i] == '0') i++;
        ans=ans.slice(i);
        return ans==""?"0":ans;
        
};